### PR TITLE
Fix-index-select-contiguous-compile

### LINKED ISSUE #168381: torch.compile / Inductor changes torch.index_select output layout from contiguous (eager) to non-contiguous
### 🐛 Describe the bug

When running under torch.compile (Inductor backend), torch.index_select produces a non-contiguous output tensor in cases where eager mode consistently returns a contiguous tensor.

This results in layout divergence between eager and compiled execution, which can break code paths that rely on the eager behavior (e.g., passing the result to external libraries expecting dense / contiguous memory).

The discrepancy appears even for very small, simple inputs and does not require gradients or complex models.
## Minimal Reproducer
```python
import torch

def test_contiguity():
    # Eager mode
    def eager_fn(x):
        x = x.clone().detach().requires_grad_(True)
        x = x.permute(1, 2, 0)       # (6,3,6) -> (3,6,6), non-contiguous
        x = x.unsqueeze(2)            # (3,6,6) -> (3,6,1,6), non-contiguous
        x = torch.relu(x)             # Still non-contiguous
        y = torch.randn(3, 6, 1, 6, device=x.device)
        x = x + y                     # Still non-contiguous
        idx = torch.tensor([0], device=x.device)
        x = torch.index_select(x, 0, idx)  # Should make contiguous
        idx = torch.tensor([0], device=x.device)
        x = torch.index_select(x, 0, idx)  # Should stay contiguous
        return x

    # Compiled mode
    compiled_fn = torch.compile(eager_fn)

    # Test
    torch.manual_seed(42)
    x1 = torch.randn(6, 3, 6)
    torch.manual_seed(42)
    x2 = torch.randn(6, 3, 6)

    eager_out = eager_fn(x1)
    compiled_out = compiled_fn(x2)

    print("Eager mode:")
    print(f"  Shape: {eager_out.shape}")
    print(f"  Stride: {eager_out.stride()}")
    print(f"  Is contiguous: {eager_out.is_contiguous()}")

    print("\nCompiled mode:")
    print(f"  Shape: {compiled_out.shape}")
    print(f"  Stride: {compiled_out.stride()}")
    print(f"  Is contiguous: {compiled_out.is_contiguous()}")

    print(f"\n❌ Contiguity mismatch: {eager_out.is_contiguous()} != {compiled_out.is_contiguous()}")
    print(f"✅ Values match: {torch.allclose(eager_out, compiled_out, atol=1e-4, rtol=1e-4)}")

    return eager_out.is_contiguous(), compiled_out.is_contiguous()

if __name__ == "__main__":
    eager_contig, compiled_contig = test_contiguity()
```
## Observed Output
```
Eager mode:
  Shape: torch.Size([1, 6, 1, 6])
  Stride: (36, 6, 6, 1)
  Is contiguous: True

Compiled mode:
  Shape: torch.Size([1, 6, 1, 6])
  Stride: (36, 1, 36, 6)
  Is contiguous: False

❌ Contiguity mismatch: True != False
✅ Values match: False
```
## Expected Behavior
torch.compile should match eager behavior unless layout semantics are explicitly relaxed.

Since eager always returns a contiguous tensor for index_select (even when the input is non-contiguous), compiled mode should match that behavior for consistency and backward compatibility.

### Versions

Collecting environment information...
PyTorch version: 2.9.1+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.13.9 | packaged by Anaconda, Inc. | (main, Oct 21 2025, 19:16:10) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.14.0-35-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.9.86
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 5090
Nvidia driver version: 575.57.08
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  24
On-line CPU(s) list:                     0-23
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Core(TM) Ultra 9 285K
CPU family:                              6
Model:                                   198
Thread(s) per core:                      1
Core(s) per socket:                      24
Socket(s):                               1
Stepping:                                2
CPU(s) scaling MHz:                      41%
CPU max MHz:                             6500.0000
CPU min MHz:                             800.0000
BogoMIPS:                                7372.80
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni lam wbnoinvd dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid bus_lock_detect movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               768 KiB (20 instances)
L1i cache:                               1.3 MiB (20 instances)
L2 cache:                                40 MiB (12 instances)
L3 cache:                                36 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-23
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS Not affected; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==2.3.5
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] torch==2.9.1
[pip3] triton==3.5.1
[conda] Could not collect

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @muchulee8 @amjames @aakhundov @coconutruben @bdhirsh

### PR DESCRIPTION
**Pull request by parsshar-RH**
Fixes #168381

### Summary:

- Fix `index_select` decomposition to return contiguous tensor

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela

### CODE REVIEW

File: test/dynamo/test_repros.py [hunk 1]
```diff
@@ -1145,6 +1145,23 @@ def f(x, y):
         self.assertEqual(out_ref, out_test)
         self.assertEqual(y_ref, y_test)
 
+    def test_index_select_contiguous_with_compile(self):
```
Comment (Lucaskabela) [inline test/dynamo/test_repros.py:1149]: nit: link the issue here for more context in a comment similar to other tests

**Review by Lucaskabela — APPROVED** [review]
Minor nit, but otherwise LGTM

File: test/dynamo/test_repros.py (hunk 1 above)
Comment (parsshar-RH) [reply test/dynamo/test_repros.py:1149]: @Lucaskabela 
i have linked the issue similar to other tests

**Comment by parsshar-RH** [comment]
@Lucaskabela

I have added issue link in the tests.
Please review!

Thanks

**Comment by parsshar-RH** [comment]
Hi @Lucaskabela @williamwen42 @ezyang 

Please review and merge the PR
Thanks

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -4311,7 +4311,7 @@ def index_select(x: TensorLike, dim: int, index: TensorLike):
         return torch.empty_like(x).index_copy(0, index, x.expand_as(index))
 
     idx = (slice(None),) * dim + (index,)
-    return x[idx]
+    return x[idx].contiguous()
```
Comment (williamwen42) [inline torch/_refs/__init__.py:4314]: Is the result guaranteed to be contiguous? Or does it depend on the input? Do you have a link to any reference?

Comment (parsshar-RH) [reply torch/_refs/__init__.py:4314]: Hi @williamwen42,
The native index_select implementations allocate the output tensor using `at::empty()` which creates a contiguous output and it does not depend in the input layout.
https://github.com/pytorch/pytorch/blob/08cd6a85ce18d3f2918942a9c18a4c56e28d58bb/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L1860-L1862
https://github.com/pytorch/pytorch/blob/08cd6a85ce18d3f2918942a9c18a4c56e28d58bb/aten/src/ATen/native/cuda/Indexing.cu#L1699-L1703

**Review by cyyever — APPROVED** [review]
